### PR TITLE
Add an RAII wrapper around KVO Observation

### PR DESCRIPTION
**Pull request by alexdeem**

### CODE REVIEW

File: TBCCore/TBCKVOObservation.h [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+@interface TBCKVOObservation : NSObject
+
++ (instancetype)observationWithSelector:(SEL)selector object:(NSObject * __nullable)object block:(TBCKVOObservationBlock)block;
+
+- (instancetype)init NS_UNAVAILABLE;
+- (instancetype)initWithSelector:(SEL)selector object:(NSObject * __nullable)object options:(NSKeyValueObservingOptions)options block:(TBCKVOObservationBlock)block __attribute((objc_designated_initializer));
+- (instancetype)initWithSelector:(SEL)selector object:(NSObject * __nullable)object block:(TBCKVOObservationBlock)block;
+
+- (void)removeObservation;
```
Comment (cysp) [inline TBCCore/TBCKVOObservation.h]: Not `-invalidate`?

File: TBCCore/TBCKVOObservation.m [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    NSString * _keyPath;
+    TBCKVOObservationBlock _block;
+    BOOL _observing;
+}
+
++ (instancetype)observationWithSelector:(SEL)selector object:(NSObject *)object block:(TBCKVOObservationBlock)block {
+    return [[self alloc] initWithSelector:selector object:object block:block];
+}
+
+- (instancetype)initWithSelector:(SEL)selector object:(NSObject * __nullable)object options:(NSKeyValueObservingOptions)options block:(TBCKVOObservationBlock)block {
```
Comment (cysp) [inline TBCCore/TBCKVOObservation.m]: Why `selector` over `keyPath`?

Comment (cysp) [reply TBCCore/TBCKVOObservation.m]: I'd've put object first too, but whatevs.

File: TBCCore/TBCKVOObservation.h [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+NS_ASSUME_NONNULL_BEGIN
+
+typedef void (^TBCKVOObservationBlock)(NSDictionary<NSKeyValueChangeKey,id> * change);
+
+@interface TBCKVOObservation : NSObject
+
++ (instancetype)observationWithSelector:(SEL)selector object:(NSObject * __nullable)object block:(TBCKVOObservationBlock)block;
+
+- (instancetype)init NS_UNAVAILABLE;
+- (instancetype)initWithSelector:(SEL)selector object:(NSObject * __nullable)object options:(NSKeyValueObservingOptions)options block:(TBCKVOObservationBlock)block __attribute((objc_designated_initializer));
```
Comment (cysp) [inline TBCCore/TBCKVOObservation.h]: `object` probably shouldn't be nullable?

File: TBCCore/TBCKVOObservation.m (hunk 2 above)
Comment (alexdeem) [reply TBCCore/TBCKVOObservation.m]: Internally, we've adopted using NSStringFromSelector(@selector(...)) when KVOing, giving the compiler the chance to complain when you make an error, or refactor.
Given that this is generic perhaps it's best to provide both.

File: TBCCore/TBCKVOObservation.h (hunk 1 above)
Comment (alexdeem) [reply TBCCore/TBCKVOObservation.h]: Consistency with `TBCNotificationCenterObservation`

File: TBCCore/TBCKVOObservation.h (hunk 3 above)
Comment (alexdeem) [reply TBCCore/TBCKVOObservation.h]: oops!

**Comment by tanyamaddalena** [comment]
👍

**Pull request merged**